File: Python/OneLang/One/Transforms/ResolveIdentifiers.py
from OneLangStdLib import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Types as types
import OneLang.One.ErrorManager as errorMan
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Statements as stats
import OneLang.One.Ast.AstTypes as astTypes

class SymbolLookup:
    def __init__(self):
        self.error_man = errorMan.ErrorManager()
        self.level_stack = []
        self.level_names = []
        self.curr_level = None
        self.symbols = Map()
    
    def throw(self, msg):
        self.error_man.throw(f'''{msg} (context: {" > ".join(self.level_names)})''')
    
    def push_context(self, name):
        self.level_stack.append(self.curr_level)
        self.level_names.append(name)
        self.curr_level = []
    
    def add_symbol(self, name, ref):
        if self.symbols.has(name):
            self.throw(f'''Symbol shadowing: {name}''')
        self.symbols.set(name, ref)
        self.curr_level.append(name)
    
    def pop_context(self):
        for name in self.curr_level:
            self.symbols.delete(name)
        self.level_names.pop()
        self.curr_level = self.level_stack.pop() if len(self.level_stack) > 0 else None
    
    def get_symbol(self, name):
        return self.symbols.get(name)

class ResolveIdentifiers(astTrans.AstTransformer):
    def __init__(self):
        self.symbol_lookup = None
        super().__init__("ResolveIdentifiers")
        self.symbol_lookup = SymbolLookup()
    
    def visit_identifier(self, id):
        super().visit_identifier(id)
        symbol = self.symbol_lookup.get_symbol(id.text)
        if symbol == None:
            self.error_man.throw(f'''Identifier \'{id.text}\' was not found in available symbols''')
            return id
        
        ref = None
        if isinstance(symbol, types.Class) and id.text == "this":
            within_static_method = isinstance(self.current_method, types.Method) and self.current_method.is_static
            ref = refs.StaticThisReference(symbol) if within_static_method else refs.ThisReference(symbol)
        elif isinstance(symbol, types.Class) and id.text == "super":
            ref = refs.SuperReference(symbol)
        else:
            ref = symbol.create_reference()
            if ref == None:
                self.error_man.throw("createReference() should not return null!")
        ref.parent_node = id.parent_node
        return ref
    
    def visit_statement(self, stmt):
        if isinstance(stmt, stats.ForStatement):
            self.symbol_lookup.push_context(f'''For''')
            if stmt.item_var != None:
                self.symbol_lookup.add_symbol(stmt.item_var.name, stmt.item_var)
            super().visit_statement(stmt)
            self.symbol_lookup.pop_context()
        elif isinstance(stmt, stats.ForeachStatement):
            self.symbol_lookup.push_context(f'''Foreach''')
            self.symbol_lookup.add_symbol(stmt.item_var.name, stmt.item_var)
            super().visit_statement(stmt)
            self.symbol_lookup.pop_context()
        elif isinstance(stmt, stats.TryStatement):
            self.symbol_lookup.push_context(f'''Try''')
            self.visit_block(stmt.try_body)
            if stmt.catch_body != None:
                self.symbol_lookup.add_symbol(stmt.catch_var.name, stmt.catch_var)
                self.visit_block(stmt.catch_body)
                self.symbol_lookup.pop_context()
            if stmt.finally_body != None:
                self.visit_block(stmt.finally_body)
        else:
            return super().visit_statement(stmt)
        return stmt
    
    def visit_lambda(self, lambda_):
        self.symbol_lookup.push_context(f'''Lambda''')
        for param in lambda_.parameters:
            self.symbol_lookup.add_symbol(param.name, param)
        super().visit_block(lambda_.body)
        # directly process method's body without opening a new scope again
        self.symbol_lookup.pop_context()
        return lambda_
    
    def visit_block(self, block):
        self.symbol_lookup.push_context("block")
        super().visit_block(block)
        self.symbol_lookup.pop_context()
        return block
    
    def visit_variable_declaration(self, stmt):
        self.symbol_lookup.add_symbol(stmt.name, stmt)
        return super().visit_variable_declaration(stmt)
    
    def visit_method_base(self, method):
        self.symbol_lookup.push_context(f'''Method: {method.name}''' if isinstance(method, types.Method) else "constructor" if isinstance(method, types.Constructor) else "???")
        
        for param in method.parameters:
            self.symbol_lookup.add_symbol(param.name, param)
            if param.initializer != None:
                self.visit_expression(param.initializer)
        
        if method.body != None:
            super().visit_block(method.body)
        # directly process method's body without opening a new scope again
        
        self.symbol_lookup.pop_context()
    
    def visit_class(self, cls_):
        self.symbol_lookup.push_context(f'''Class: {cls_.name}''')
        self.symbol_lookup.add_symbol("this", cls_)
        if isinstance(cls_.base_class, astTypes.ClassType):
            self.symbol_lookup.add_symbol("super", cls_.base_class.decl)
        super().visit_class(cls_)
        self.symbol_lookup.pop_context()
    
    def visit_file(self, source_file):
        self.error_man.reset_context(self)
        self.symbol_lookup.push_context(f'''File: {source_file.source_path.to_string()}''')
        
        for symbol in source_file.available_symbols.values():
            if isinstance(symbol, types.Class):
                self.symbol_lookup.add_symbol(symbol.name, symbol)
            elif isinstance(symbol, types.Interface):
                pass
            elif isinstance(symbol, types.Enum):
                self.symbol_lookup.add_symbol(symbol.name, symbol)
            elif isinstance(symbol, types.GlobalFunction):
                self.symbol_lookup.add_symbol(symbol.name, symbol)
            else:
                pass
        
        super().visit_file(source_file)
        
        self.symbol_lookup.pop_context()
        self.error_man.reset_context()

File: PHP/src/OneLang/One/Transforms/ResolveIdentifiers.php
<?php

namespace OneLang\One\Transforms\ResolveIdentifiers;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Enum;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Types\Lambda;
use OneLang\One\Ast\Types\GlobalFunction;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\Types\Constructor;
use OneLang\One\Ast\Types\Interface_;
use OneLang\One\ErrorManager\ErrorManager;
use OneLang\One\Ast\Expressions\Identifier;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\References\IReferencable;
use OneLang\One\Ast\References\Reference;
use OneLang\One\Ast\References\StaticThisReference;
use OneLang\One\Ast\References\ThisReference;
use OneLang\One\Ast\References\SuperReference;
use OneLang\One\Ast\Statements\VariableDeclaration;
use OneLang\One\Ast\Statements\ForStatement;
use OneLang\One\Ast\Statements\ForeachStatement;
use OneLang\One\Ast\Statements\Statement;
use OneLang\One\Ast\Statements\IfStatement;
use OneLang\One\Ast\Statements\TryStatement;
use OneLang\One\Ast\Statements\Block;
use OneLang\One\Ast\AstTypes\ClassType;

class SymbolLookup {
    public $errorMan;
    public $levelStack;
    public $levelNames;
    public $currLevel;
    public $symbols;
    
    function __construct()
    {
        $this->errorMan = new ErrorManager();
        $this->levelStack = array();
        $this->levelNames = array();
        $this->symbols = new \OneLang\Core\Map();
    }
    
    function throw($msg) {
        $this->errorMan->throw($msg . " (context: " . implode(" > ", $this->levelNames) . ")");
    }
    
    function pushContext($name) {
        $this->levelStack[] = $this->currLevel;
        $this->levelNames[] = $name;
        $this->currLevel = array();
    }
    
    function addSymbol($name, $ref) {
        if ($this->symbols->has($name))
            $this->throw("Symbol shadowing: " . $name);
        $this->symbols->set($name, $ref);
        $this->currLevel[] = $name;
    }
    
    function popContext() {
        foreach ($this->currLevel as $name)
            $this->symbols->delete($name);
        array_pop($this->levelNames);
        $this->currLevel = count($this->levelStack) > 0 ? array_pop($this->levelStack) : null;
    }
    
    function getSymbol($name) {
        return $this->symbols->get($name);
    }
}

class ResolveIdentifiers extends AstTransformer {
    public $symbolLookup;
    
    function __construct() {
        parent::__construct("ResolveIdentifiers");
        $this->symbolLookup = new SymbolLookup();
    }
    
    protected function visitIdentifier($id) {
        parent::visitIdentifier($id);
        $symbol = $this->symbolLookup->getSymbol($id->text);
        if ($symbol === null) {
            $this->errorMan->throw("Identifier '" . $id->text . "' was not found in available symbols");
            return $id;
        }
        
        $ref = null;
        if ($symbol instanceof Class_ && $id->text === "this") {
            $withinStaticMethod = $this->currentMethod instanceof Method && $this->currentMethod->isStatic;
            $ref = $withinStaticMethod ? new StaticThisReference($symbol) : new ThisReference($symbol);
        }
        else if ($symbol instanceof Class_ && $id->text === "super")
            $ref = new SuperReference($symbol);
        else {
            $ref = $symbol->createReference();
            if ($ref === null)
                $this->errorMan->throw("createReference() should not return null!");
        }
        $ref->parentNode = $id->parentNode;
        return $ref;
    }
    
    protected function visitStatement($stmt) {
        if ($stmt instanceof ForStatement) {
            $this->symbolLookup->pushContext("For");
            if ($stmt->itemVar !== null)
                $this->symbolLookup->addSymbol($stmt->itemVar->name, $stmt->itemVar);
            parent::visitStatement($stmt);
            $this->symbolLookup->popContext();
        }
        else if ($stmt instanceof ForeachStatement) {
            $this->symbolLookup->pushContext("Foreach");
            $this->symbolLookup->addSymbol($stmt->itemVar->name, $stmt->itemVar);
            parent::visitStatement($stmt);
            $this->symbolLookup->popContext();
        }
        else if ($stmt instanceof TryStatement) {
            $this->symbolLookup->pushContext("Try");
            $this->visitBlock($stmt->tryBody);
            if ($stmt->catchBody !== null) {
                $this->symbolLookup->addSymbol($stmt->catchVar->name, $stmt->catchVar);
                $this->visitBlock($stmt->catchBody);
                $this->symbolLookup->popContext();
            }
            if ($stmt->finallyBody !== null)
                $this->visitBlock($stmt->finallyBody);
        }
        else
            return parent::visitStatement($stmt);
        return $stmt;
    }
    
    protected function visitLambda($lambda) {
        $this->symbolLookup->pushContext("Lambda");
        foreach ($lambda->parameters as $param)
            $this->symbolLookup->addSymbol($param->name, $param);
        parent::visitBlock($lambda->body);
        // directly process method's body without opening a new scope again
        $this->symbolLookup->popContext();
        return $lambda;
    }
    
    protected function visitBlock($block) {
        $this->symbolLookup->pushContext("block");
        parent::visitBlock($block);
        $this->symbolLookup->popContext();
        return $block;
    }
    
    protected function visitVariableDeclaration($stmt) {
        $this->symbolLookup->addSymbol($stmt->name, $stmt);
        return parent::visitVariableDeclaration($stmt);
    }
    
    protected function visitMethodBase($method) {
        $this->symbolLookup->pushContext($method instanceof Method ? "Method: " . $method->name : ($method instanceof Constructor ? "constructor" : "???"));
        
        foreach ($method->parameters as $param) {
            $this->symbolLookup->addSymbol($param->name, $param);
            if ($param->initializer !== null)
                $this->visitExpression($param->initializer);
        }
        
        if ($method->body !== null)
            parent::visitBlock($method->body);
        // directly process method's body without opening a new scope again
        
        $this->symbolLookup->popContext();
    }
    
    protected function visitClass($cls) {
        $this->symbolLookup->pushContext("Class: " . $cls->name);
        $this->symbolLookup->addSymbol("this", $cls);
        if ($cls->baseClass instanceof ClassType)
            $this->symbolLookup->addSymbol("super", $cls->baseClass->decl);
        parent::visitClass($cls);
        $this->symbolLookup->popContext();
    }
    
    function visitFile($sourceFile) {
        $this->errorMan->resetContext($this);
        $this->symbolLookup->pushContext("File: " . $sourceFile->sourcePath->toString());
        
        foreach ($sourceFile->availableSymbols->values() as $symbol) {
            if ($symbol instanceof Class_)
                $this->symbolLookup->addSymbol($symbol->name, $symbol);
            else if ($symbol instanceof Interface_) { }
            else if ($symbol instanceof Enum)
                $this->symbolLookup->addSymbol($symbol->name, $symbol);
            else if ($symbol instanceof GlobalFunction)
                $this->symbolLookup->addSymbol($symbol->name, $symbol);
            else { }
        }
        
        parent::visitFile($sourceFile);
        
        $this->symbolLookup->popContext();
        $this->errorMan->resetContext();
    }
}
